sql: fix race in app stats

RaduBerinde opened the pull request:
I introduced a race by retrieving the distsql/vectorized flags from the
statement entry without locking it. This change fixes it by moving the access to
the existing locked section.

Fixes #53000.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/53005)
<!-- Reviewable:end -->

RaduBerinde commented:
Reproduced on master with stressrace and verified that it's fixed.

RaduBerinde commented:
bors r+

RaduBerinde committed:
sql: fix race in app stats

I introduced a race by retrieving the distsql/vectorized flags from the
statement entry without locking it. This change fixes it by moving the access to
the existing locked section.

Release note: None

jbowens commented:
@RaduBerinde is this good to merge?

RaduBerinde commented:
Bors was supposed to merge it..

bors r+
